docs/consistent-hash.md

```diff
@@ -0,0 +1,91 @@
+
+For all data $k \in U_k$, where $U_k$ is an unbounded set, we apply a hash function $v = H(k)$, where $v$ falls to a limited range. The hash function $H$ ensures that all data are hashed **uniformly** to that range. We call $v$ vnode, namely virtual node, as is shown as the squares in the figure below.
+
+![initial data distribution](./images/consistent-hash/data-distribution.svg)
+
+Then we have vnode mapping, which ensures that vnodes are mapped evenly to parallel units in the cluster. In other words, the number of vnodes that are mapped to each parallel unit should be as close as possible. This is denoted by different colors in the figure above. As is depicted, we have 3 parallel units (shown as circles), each taking $\frac{1}{3}$ of total vnodes. Vnode mapping is [constructed and maintained by meta](https://github.com/singularity-data/risingwave/blob/main/src/meta/src/manager/hash_mapping.rs).
+
+As long as the hash function $H$ could ensure uniformity, the data distribution determined by this strategy would be even across physical resources, even if data in $U_k$ skew to a certain range. 
+
+#### Data Redistribution
+
+Since $v = H(k)$, the way that data are mapped to vnodes will be invariant. Therefore, when scaling occurs, we only need to modify vnode mapping (the way that vnodes are mapped to parallel units), so as to redistribute the data.
+
+Let's take scaling out for example. Assume that we have one more parallel unit after scaling out, as is depicted as the orange circle in the figure below. Using the optimal strategy, we modify the vnode mapping in such a way that only $\frac{1}{4}$ of the data have to be moved, as is shown in the figure below. The vnodes whose data are required to be moved are highlighted with bold border in the figure.
+
+![optimal data redistribution](./images/consistent-hash/data-redistribution-1.svg)
+
+To minimize data movement when scaling occurs, we should be careful when we modify the vnode mapping. Below is an opposite example: modifying vnode mapping like this will result in $\frac{1}{2}$ of the data being moved.
+
+![worst data redistribution](./images/consistent-hash/data-redistribution-2.svg)
+
+### Streaming
+
+We know that a fragment has several actors as its different parallelisms, and that upstream actors will send data to downstream actors via [dispatcher](./streaming-overview.md#actors). The figure below illustrates how actors distribute data based on consistent hash by example. 
+
+![actor data distribution](./images/consistent-hash/actor-data.svg)
+
+In the figure, we can see that one upstream actor dispatches data to three downstream actors. The downstream actors are scheduled on the parallel units mentioned in previous example respectively. 
+
+Based on our consistent hash design, the dispatcher is informed of the latest vnode mapping by meta node. It then decides how to send data by following steps:
```

Comment:
Is it possible that vnode mapping in CN is stale? How meta inform the latest to CN?

Reply:
This is, in fact, not implemented yet. When scaling occurs, actors are designed to be dropped and recreated (currently), so the dispatchers would know the latest vnode mapping accordingly.